### PR TITLE
Bootstrap API refactoring

### PR DESCRIPTION
**Pull request by aloubyansky**
This PR introduces a common Quarkus application model and workspace API that are supposed to be used for all the supported build systems (e.g. Maven, Gradle, JBang, etc).
I still plan to evolve the representation of the workspace module sources but I was thinking I'd open this PR now for review, since it's already a huge change.
It also reduces the number of workspace and application model resolutions for the dev mode. I am looking into getting it down to a single app model initialization in a different branch.

### CODE REVIEW

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/model/ApplicationModel.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    Set<ArtifactKey> getReloadableWorkspaceDependencies();
+
+    default WorkspaceModule getApplicationModule() {
+        return getAppArtifact().getWorkspaceModule();
+    }
+
+    default Collection<WorkspaceModule> getWorkspaceModules() {
+        final List<WorkspaceModule> modules = new ArrayList<>();
+        //ProjectModule module = getApplicationModule();
```
Comment (glefloch) [inline independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/model/ApplicationModel.java:53]: Should we keep this commented block of code?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/jbang/JBangBuilderImpl.java [hunk 2]
```diff
@@ -46,28 +49,28 @@
 
         try {
             Path target = Files.createTempDirectory("quarkus-jbang");
-            AppArtifact appArtifact = new AppArtifact("dev.jbang.user", "quarkus", null, "jar", "999-SNAPSHOT");
-            appArtifact.setPath(appClasses);
+            final ResolvedArtifactDependency appArtifact = new ResolvedArtifactDependency("dev.jbang.user", "quarkus", null,
+                    "jar", "999-SNAPSHOT", appClasses);
```
Comment (glefloch) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/jbang/JBangBuilderImpl.java:53]: It was present before your modification but should we get rid of the hard coded version?

**Review by glefloch — COMMENTED** [review]
Nice work! That will definitely ease application model usage. I left few comments.

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java [hunk 3]
```diff
@@ -13,74 +14,85 @@
...
-    public void relink(AppArtifact appArtifact, Path localPath) throws AppModelResolverException {
+    public void relink(ArtifactCoords artifact, Path localPath)
+            throws AppModelResolverException {
 
     }
 
     @Override
-    public Path resolve(AppArtifact appArtifact) throws AppModelResolverException {
-        return resolveArtifact(appArtifact).getPaths().getSinglePath();
+    public io.quarkus.maven.dependency.ResolvedDependency resolve(ArtifactCoords appArtifact) throws AppModelResolverException {
```
[the remark this answers is not in the document]
Comment (aloubyansky) [reply devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java:85]: Right, now the class/interface names don't conflict with the Gradle API, adding the imports instead.

File: devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/extension/QuarkusPluginExtension.java [hunk 4]
```diff
@@ -249,7 +238,8 @@ private File getLastFile(FileCollection fileCollection) {
      * @return the source sets associated with the current project.
      */
     private SourceSetContainer getSourceSets() {
-        return project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets();
+        final JavaPluginExtension javaExtension = project.getExtensions().findByType(JavaPluginExtension.class);
```
[the remark this answers is not in the document]
Comment (aloubyansky) [reply devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/extension/QuarkusPluginExtension.java:241]: Ok, changing back to the convention, thanks!

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/model/ApplicationModel.java (hunk 1 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/model/ApplicationModel.java:53]: Originally I left it as a reminder to myself to review why we don't add the application module to the complete list of locally available dependencies. But it's not a big deal. I'll remove it.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/model/ApplicationModel.java:53]: Originally I left it as a reminder to myself to review why we don't add the application module to the complete list of locally available dependencies. But it's not a big deal. I'll remove it.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/jbang/JBangBuilderImpl.java (hunk 2 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/jbang/JBangBuilderImpl.java:53]: I think it's not important what this version actually is. I'll leave it for now.

**Comment by aloubyansky** [comment]
@stuartwdouglas what do you think about getting this in 2.4?

**Review by stuartwdouglas — APPROVED** [review]
I think this looks good.

**Pull request merged**